### PR TITLE
Add health check api in admin handler

### PR DESCRIPTION
**Pull request by yux0**
## What changed?
Add health check in admin handler

## Why?
Add Health check API in admin handler to support check the cluster health status

### CODE REVIEW

File: service/frontend/health_check.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	for _, client := range c.clients {
+		go func(hc healthpb.HealthClient) {
+			resp, err := hc.Check(ctx, &healthpb.HealthCheckRequest{Service: c.serviceName})
+			respCh <- healthCheckResponse{
+				HealthCheckResponse: resp,
+				err:                 err,
+			}
+		}(client)
+	}
+	close(respCh)
```
Comment (hai719) [inline service/frontend/health_check.go:49]: should use waitgroup to wait hc.Check to finish before close the channel.

File: service/frontend/health_check.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	var failureCount float64
+	for resp := range respCh {
+		if resp.err != nil || resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
```
Comment (hai719) [inline service/frontend/health_check.go:53]: Should we distinguish "failed to check" from "not serving"?

Comment (yux0) [reply service/frontend/health_check.go:53]: what is the consideration? For those two, it only checks the connection.

Comment (hai719) [reply service/frontend/health_check.go:53]: Ok if it only checks connection.

**Review by hai719 — APPROVED** [review]

File: cmd/tools/rpcwrappers/main.go [hunk 3]
```diff
@@ -96,6 +97,7 @@ var (
 		"client.history.StreamWorkflowReplicationMessages":          true,
 		"metricsClient.history.StreamWorkflowReplicationMessages":   true,
 		"retryableClient.history.StreamWorkflowReplicationMessages": true,
+		"client.admin.HealthCheck":                                  true,
```
Comment (dnr) [inline cmd/tools/rpcwrappers/main.go:100]: why does this have to be excluded? the generated one looks fine

File: common/dynamicconfig/constants.go [hunk 4]
```diff
@@ -967,6 +967,16 @@ Default is 4.`,
 after all namespace resources (i.e. workflow executions) are deleted.
 Default is 0, means, namespace will be deleted immediately.`,
 	)
+	FrontendHealthCheckFailureRate = NewGlobalFloatSetting(
+		"frontend.healthCheckFailureRat",
```
Comment (dnr) [inline common/dynamicconfig/constants.go:971]:
```suggestion
		"frontend.healthCheckFailureRate",
```

File: common/dynamicconfig/constants.go [hunk 5]
```diff
@@ -967,6 +967,16 @@ Default is 4.`,
...
+		0.8,
+		`FrontendServiceHealthCheckFailureRate is the percentile of failed hosts on frontend health check`,
```
Comment (dnr) [inline common/dynamicconfig/constants.go:973]: percentile of failed hosts that do what? explain the semantics clearly

also I don't think it's a percentile, it's a percentage

File: cmd/tools/rpcwrappers/main.go [hunk 6]
```diff
@@ -105,12 +107,14 @@ var (
 		"client.history.DeleteDLQTasks":         true,
 		"client.history.ListQueues":             true,
 		"client.history.ListTasks":              true,
+		"client.history.HealthCheck":            true,
```
Comment (dnr) [inline cmd/tools/rpcwrappers/main.go:110]: why is this here? there's no call by this name in HistoryService (or MatchingService below)

File: service/frontend/health_check.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+func NewHealthCheckClientFactory(
+	membershipMonitor membership.Monitor,
+	rpcFactory common.RPCFactory,
+	config *Config,
+) HealthCheckClientFactory {
+	factory := HealthCheckClientFactory{
+		membershipMonitor: membershipMonitor,
+		rpcFactory:        rpcFactory,
+		config:            config,
+		connectionCache:   cache.NewLRU(500, nil),
```
Comment (dnr) [inline service/frontend/health_check.go:98]: :grimacing:   why does this have its own connection cache? it should share a cache with the existing clients

File: proto/internal/temporal/server/api/adminservice/v1/request_response.proto [hunk 8]
```diff
@@ -553,3 +553,11 @@ message ListQueuesResponse {
   repeated QueueInfo queues = 1;
   bytes next_page_token = 2;
 }
+
+message HealthCheckRequest {
```
Comment (dnr) [inline proto/internal/temporal/server/api/adminservice/v1/request_response.proto:557]: My main comment: please don't call this a "health check". We already have a thing called a health check, the standard grpc health check, with a health server, etc. Calling this (which is very different) a health check is very confusing.

Please rename everything to DeepHealthCheck or DependencyHealthCheck or something like that.

File: service/frontend/service.go [hunk 9]
```diff
@@ -308,6 +312,9 @@ func NewConfig(
 		AdminEnableListHistoryTasks: dynamicconfig.AdminEnableListHistoryTasks.Get(dc),
 
 		MaskInternalErrorDetails: dynamicconfig.FrontendMaskInternalErrorDetails.Get(dc),
+
+		FrontendHealthCheckFailureRate: dynamicconfig.FrontendHealthCheckFailureRate.Get(dc),
+		MatchingHealthCheckFailureRate: dynamicconfig.MatchingHealthCheckFailureRate.Get(dc),
```
Comment (yycptt) [inline service/frontend/service.go:317]: Not sure those are data plane config level, I was thinking server just return the data/stats and control plane will make the decision. It probably not a big deal if it's just those two configs but I guess the deep health check logic will become more and more complicate.

File: service/frontend/health_check.go (hunk 7 above)
Comment (yycptt) [reply service/frontend/health_check.go:98]: I guess that requires some refactoring in the rpcFactory impl, but yeah we should reuse the connection.

File: service/frontend/admin_handler.go [hunk 10]
```diff
@@ -238,6 +245,36 @@ func (adh *AdminHandler) Stop() {
 	}
 }
 
+func (adh *AdminHandler) HealthCheck(
```
Comment (yycptt) [inline service/frontend/admin_handler.go:248]: you may want to exclude health check endpoint from rate limiting or make it high priority.

File: cmd/tools/rpcwrappers/main.go (hunk 3 above)
Comment (yux0) [reply cmd/tools/rpcwrappers/main.go:100]: removed. use the generated one.

File: common/dynamicconfig/constants.go (hunk 4 above)
Comment (yux0) [reply common/dynamicconfig/constants.go:971]: updated

File: common/dynamicconfig/constants.go (hunk 5 above)
Comment (yux0) [reply common/dynamicconfig/constants.go:973]: updated

File: service/frontend/health_check.go (hunk 7 above)
Comment (yux0) [reply service/frontend/health_check.go:98]: I guess I need some refactor here as I was trying to use the grpc health check api. But I think I need to introduce deep health check to both matching and frontend.

File: proto/internal/temporal/server/api/adminservice/v1/request_response.proto (hunk 8 above)
Comment (yux0) [reply proto/internal/temporal/server/api/adminservice/v1/request_response.proto:557]: Changed to deep health check

File: service/frontend/service.go (hunk 9 above)
Comment (yux0) [reply service/frontend/service.go:317]: We should sync offline on this. I think the cluster to tell if itself is healthy. With return stats, this api is similar to metrics and we might updating the response frequent.

File: service/frontend/admin_handler.go (hunk 10 above)
Comment (yux0) [reply service/frontend/admin_handler.go:248]: good catch.

File: service/frontend/health_check.go (hunk 7 above)
Comment (yux0) [reply service/frontend/health_check.go:98]: I think this requires more change. I will remove this for now and move to a different PR.

File: service/frontend/fx.go [hunk 11]
```diff
@@ -605,6 +607,7 @@ func AdminHandlerProvider(
 	eventSerializer serialization.Serializer,
 	timeSource clock.TimeSource,
 	taskCategoryRegistry tasks.TaskCategoryRegistry,
+	rpcFactory common.RPCFactory,
```
Comment (yycptt) [inline service/frontend/fx.go:610]: ditto

File: service/frontend/admin_handler.go [hunk 12]
```diff
@@ -127,6 +127,7 @@ type (
 		saManager                  searchattribute.Manager
 		clusterMetadata            cluster.Metadata
 		healthServer               *health.Server
+		healthCheckClientFactory   HealthCheckClientFactory
```
Comment (yycptt) [inline service/frontend/admin_handler.go:130]: ditto

File: service/frontend/admin_handler.go [hunk 13]
```diff
@@ -159,6 +160,7 @@ type (
 		HealthServer                        *health.Server
 		EventSerializer                     serialization.Serializer
 		TimeSource                          clock.TimeSource
+		RpcFactory                          common.RPCFactory
```
Comment (yycptt) [inline service/frontend/admin_handler.go:163]: ditto

**Review by yycptt — APPROVED** [review]

File: cmd/tools/rpcwrappers/main.go [hunk 14]
```diff
@@ -111,6 +112,7 @@ var (
 		"client.matching.PollActivityTaskQueue": true,
 		"client.matching.PollWorkflowTaskQueue": true,
 		"client.matching.QueryWorkflow":         true,
+		"client.matching.HealthCheck":           true,
```
[the remark this answers is not in the document]
Comment (yux0) [reply cmd/tools/rpcwrappers/main.go:115]: cleanded up

File: service/frontend/admin_handler.go (hunk 12 above)
Comment (yux0) [reply service/frontend/admin_handler.go:130]: cleaned up

File: service/frontend/admin_handler.go (hunk 13 above)
Comment (yux0) [reply service/frontend/admin_handler.go:163]: cleaned up
